subprojects/platform-native/src/main/java/org/gradle/nativeplatform/toolchain/internal/AbstractCompiler.java

```diff
@@ -34,18 +35,27 @@
     private final CommandLineToolContext invocationContext;
     private final boolean useCommandFile;
     private final BuildOperationExecutor buildOperationExecutor;
+    private final WorkerLeaseService workerLeaseService;
 
-    protected AbstractCompiler(BuildOperationExecutor buildOperationExecutor, CommandLineToolInvocationWorker commandLineToolInvocationWorker, CommandLineToolContext invocationContext, ArgsTransformer<T> argsTransformer, boolean useCommandFile) {
+    protected AbstractCompiler(BuildOperationExecutor buildOperationExecutor, CommandLineToolInvocationWorker commandLineToolInvocationWorker, CommandLineToolContext invocationContext, ArgsTransformer<T> argsTransformer, boolean useCommandFile, WorkerLeaseService workerLeaseService) {
         this.buildOperationExecutor = buildOperationExecutor;
         this.argsTransformer = argsTransformer;
         this.invocationContext = invocationContext;
         this.useCommandFile = useCommandFile;
         this.commandLineToolInvocationWorker = commandLineToolInvocationWorker;
+        this.workerLeaseService = workerLeaseService;
     }
 
     @Override
-    public WorkResult execute(T spec) {
-        buildOperationExecutor.runAll(commandLineToolInvocationWorker, newInvocationAction(spec));
+    public WorkResult execute(final T spec) {
+        final Action<BuildOperationQueue<CommandLineToolInvocation>> invocationAction = newInvocationAction(spec);
+
+        workerLeaseService.withoutProjectLock(new Runnable() {
+            @Override
+            public void run() {
+                buildOperationExecutor.runAll(commandLineToolInvocationWorker, invocationAction);
```

Comment:
Ah, yes - I misunderstood what you were referring to, and of course, you are right.  I pulled it up so that the args passed in to `newInvocation()` as you've suggested.